Add primitive routines to perform Audio Sample format conversion.

jyavenard opened the pull request:
#### bf54b5a538ebd488a067445fef3a72c7067051bc
<pre>
Add primitive routines to perform Audio Sample format conversion.
<a href="https://bugs.webkit.org/show_bug.cgi?id=283902">https://bugs.webkit.org/show_bug.cgi?id=283902</a>
<a href="https://rdar.apple.com/140782221">rdar://140782221</a>

Reviewed by Gerald Squelart.

Add convertAudioSample method to convert uint8_t, int16_t, int32_t, float audio sample formats.

Float Audio samples use a float range of [-1.0, 1.0] inclusive.

This method will be used with the WebCodec&apos;s AudioData API.
The conversion had two requirements to fullfill the WebCodec&apos;s AudioData wpt:
 1- The conversion is symmetric between int and float.
 2- +1.0 can be mapped to std::numeric_limits&lt;T&gt;::max().

Requirement 2) was removed in later WebCodec&apos;s wpt, but for the version currently in the tree, it is required.

Float conversion accuracy has been given preferential treatment over speed.

Add API tests.
* Source/WebCore/platform/AudioSampleFormat.h:
(WebCore::detail::maxAsFloat):
(WebCore::detail::maxPlus1AsFloat):
(WebCore::detail::lowestAsFloat):
(WebCore::detail::divideWithBiasAndClamp):
(WebCore::detail::uint8ToAudioSample):
(WebCore::detail::int16ToAudioSample):
(WebCore::detail::int32ToAudioSample):
(WebCore::detail::floatToAudioSample):
(WebCore::convertAudioSample):
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/AudioSampleFormat.cpp: Added.
(TestWebKitAPI::TEST(AudioSampleFormat, U8)):
(TestWebKitAPI::TEST(AudioSampleFormat, U8toS16)):
(TestWebKitAPI::TEST(AudioSampleFormat, S16)):
(TestWebKitAPI::TEST(AudioSampleFormat, S16toS32)):
(TestWebKitAPI::TEST(AudioSampleFormat, S32)):
(TestWebKitAPI::TEST(AudioSampleFormat, F32)):

Canonical link: <a href="https://commits.webkit.org/287327@main">https://commits.webkit.org/287327@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb64017d1a9b41f1e5aaec8e1802ba9880ce9c35

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79043 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58079 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32420 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83683 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30258 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81176 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67188 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6349 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61867 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19786 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82110 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51906 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72049 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42170 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49257 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26143 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28622 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70372 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26561 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85069 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6368 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4400 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70105 "Found 1 new test failure: webanimations/accelerated-animation-addition-lower-in-effect-stack.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6532 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67921 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69355 "Found 10 new API test failures: /TestWebKit:WebKit.DownloadDecideDestinationCrash, /TestWebKit:WebKit.DocumentStartUserScriptAlertCrashTest, /TestWebKit:WebKit.EvaluateJavaScriptThatCreatesBlob, /TestWebKit:WebKit.InjectedBundleFrameHitTest, /TestWebKit:WebKit.HitTestResultNodeHandle, /TestWebKit:WebKit.LoadCanceledNoServerRedirectCallback, /TestWebKit:WebKit.FrameMIMETypePNG, /TestWebKit:WebKit.FocusedFrameAfterCrash, /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginShouldNotBeDispatchedForAlreadyFocusedField, /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginAndEndEditingEvents (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17312 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13400 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12176 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6319 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12155 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6280 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9731 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8072 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WebCore/platform/AudioSampleFormat.h:
```diff
@@ -39,4 +42,113 @@ enum class AudioSampleFormat {
+    if constexpr (std::is_same_v<T, uint8_t>)
+        return value;
+    else if constexpr (std::is_same_v<T, int16_t>)
+        return (static_cast<int16_t>(value) << 8) - (1 << 15);
```
Nit: Due to [integral promotion](https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_promotion), I believe that the arguments of `<<` are first converted to int. Hence that `static_cast` is effectively useless!
However, when returning, the result will be implicitly converted to T (int16_t in this case), and that could trigger the "-Wconversion" warning (if we use it?). To avoid that, you should have your `static_cast` surround the entire expression.
Have a look at https://cppinsights.io/s/8ffb5ab5 , it may be useful to check the rest of your code in cppinsights, in case I've missed something.

jyavenard replied:
TIL

jyavenard replied:
I did add that cast because I got an error in the WPT test... I'll add more tests as well then

squelart commented on Source/WebCore/platform/AudioSampleFormat.h:
```diff
@@ -39,4 +42,113 @@ enum class AudioSampleFormat {
+    else if constexpr (std::is_same_v<T, uint8_t>)
+        return static_cast<T>(std::clamp((value + 1.0f) * 128.0f, lowestAsFloat<T>(), maxAsFloat<T>()));
+    else if constexpr (std::is_same_v<T, int16_t>)
+        return static_cast<T>(std::clamp(value * -lowestAsFloat<T>(), lowestAsFloat<T>(), maxAsFloat<T>()));
```
This `-lowestAsFloat<T>()` looks like magic! For clarity I'd suggest moving it to a new function called `maxPlus1AsFloat<T>()` or similar.

squelart commented on Tools/TestWebKitAPI/Tests/WebCore/AudioSampleFormat.cpp:
```diff
@@ -0,0 +1,139 @@
+    for (size_t i = 0; i < testVectorUint8.size(); i++)
+        EXPECT_EQ(testVectorUint8[i], testVectorUint8FromFloat[i]);
+
+    // Test that the min and max of the [-1, 1] range can be reached.
```
Very optional nit: Comment doesn't quite fit the next 3 lines. Maybe mention 0.0<->128, I guess it's an important value. Or just remove the comment?
And why no comment for the first sub-test above?

& same for the next tests.

squelart reviewed (approved):
Good to have these tests!
r+ with nits addressed if/as you see fit.

jyavenard committed:
Add primitive routines to perform Audio Sample format conversion.
https://bugs.webkit.org/show_bug.cgi?id=283902
rdar://140782221

Reviewed by Gerald Squelart.

Add convertAudioSample method to convert uint8_t, int16_t, int32_t, float audio sample formats.

Float Audio samples use a float range of [-1.0, 1.0] inclusive.

This method will be used with the WebCodec's AudioData API.
The conversion had two requirements to fullfill the WebCodec's AudioData wpt:
 1- The conversion is symmetric between int and float.
 2- +1.0 can be mapped to std::numeric_limits<T>::max().

Requirement 2) was removed in later WebCodec's wpt, but for the version currently in the tree, it is required.

Float conversion accuracy has been given preferential treatment over speed.

Add API tests.
* Source/WebCore/platform/AudioSampleFormat.h:
(WebCore::detail::maxAsFloat):
(WebCore::detail::maxPlus1AsFloat):
(WebCore::detail::lowestAsFloat):
(WebCore::detail::divideWithBiasAndClamp):
(WebCore::detail::uint8ToAudioSample):
(WebCore::detail::int16ToAudioSample):
(WebCore::detail::int32ToAudioSample):
(WebCore::detail::floatToAudioSample):
(WebCore::convertAudioSample):
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/AudioSampleFormat.cpp: Added.
(TestWebKitAPI::TEST(AudioSampleFormat, U8)):
(TestWebKitAPI::TEST(AudioSampleFormat, U8toS16)):
(TestWebKitAPI::TEST(AudioSampleFormat, S16)):
(TestWebKitAPI::TEST(AudioSampleFormat, S16toS32)):
(TestWebKitAPI::TEST(AudioSampleFormat, S32)):
(TestWebKitAPI::TEST(AudioSampleFormat, F32)):

Canonical link: https://commits.webkit.org/287327@main
